[NUI] Add InsetsChanged event to Window

jaehyun0cho opened the pull request:
Window insets are the inner size between window content and window edge.

Window insets are required when status bar, virtual keyboard, or clipboard appears on window to avoid window content obscured by them.

To notify the window insets changes, InsetsChanged event is added to Window.

### Description of Change ###
<!-- Describe your changes here. -->


### API Changes ###
<!-- If you have the ACR for changing APIs, put the link of the ACR: -->
 - ACR:

<!--
If you can't make the ACR, List all API changes here (or just put None), example:
Added:
 - SafeBundleHandle Bundle.SafeBundleHandle { get; } // Property
 - void Bundle.AddItem(string key, string value);

Changed:
 - object Bundle.ReceiveItem(string key) => object Bundle.GetItem(string key);
-->

jaehyun0cho commented:
This PR requires the following patches.
https://review.tizen.org/gerrit/#/c/platform/core/uifw/dali-adaptor/+/277884/
https://review.tizen.org/gerrit/#/c/platform/core/uifw/dali-csharp-binder/+/277885/

hinohie commented on src/Tizen.NUI/src/public/Window/WindowEvent.cs:
```diff
@@ -1551,6 +1562,127 @@ private void OnAuxiliaryMessage(IntPtr kData, IntPtr vData, IntPtr optionsArray)
+                        using var signal = new WindowInsetsChangedSignal(this);
+                        signal.Disconnect(insetsChangedEventCallback);
+
+                        if (signal.Empty())
```
과거어 Signal 관련 함수들 구현 정리하면서, Connect / Disconnect 이외의 모든 API 를 전부 지운 이력이 있습니다. 이 Signal 도 이에 해당되어 보입니다.

jaehyun0cho replied:
Updated the code as you suggest.
I am sorry for updating late.

hinohie commented on src/Tizen.NUI/src/public/Window/WindowEvent.cs:
```diff
@@ -1551,6 +1562,127 @@ private void OnAuxiliaryMessage(IntPtr kData, IntPtr vData, IntPtr optionsArray)
+                insetsChangedEventHandler -= value;
+                if (insetsChangedEventHandler == null)
+                {
+                    if (insetsChangedEventCallback != null)
```
```suggestion
                    if (insetsChangedEventCallback != null && insetsChangedEventCallback != null)
```

Dispose 된 Window 에다가 이벤트 -= 를 수행히려고 해서 발생한 버그가 있었습니다. ㅇ이에 대비하는 방어코드가 필요합니다

jaehyun0cho replied:
Updated the code as you suggest.
I am sorry for updating late.

dongsug-song commented:
@jaehyun0cho 
안녕하세요. hinohie 말씀대로 event 코드 패턴이 변경 되었습니다. 
예: https://github.com/Samsung/TizenFX/blob/master/src/Tizen.NUI/src/public/Window/WindowEvent.cs#L91
같은 패치인 https://github.com/Samsung/TizenFX/pull/5421 는 tizen_7.0 이라 다른 코드들과 패턴이 같지만 tizen_8.0 에서는 새로 방식을 맞출 필요가 있어 보입니다. 
확인 부탁드립니다. 감사합니다.

jaehyun0cho commented:
> @jaehyun0cho 안녕하세요. hinohie 말씀대로 event 코드 패턴이 변경 되었습니다. 예: https://github.com/Samsung/TizenFX/blob/master/src/Tizen.NUI/src/public/Window/WindowEvent.cs#L91 같은 패치인 #5421 는 tizen_7.0 이라 다른 코드들과 패턴이 같지만 tizen_8.0 에서는 새로 방식을 맞출 필요가 있어 보입니다. 확인 부탁드립니다. 감사합니다.

I updated this PR based on the latest event code pattern.
I am sorry that I updated late.

taehyub reviewed (approved):
LGTM

dongsug-song commented on src/Tizen.NUI/src/public/Window/WindowEvent.cs:
```diff
@@ -85,6 +86,8 @@ public partial class Window
         private delegate void MoveCompletedEventCallbackType(IntPtr window, IntPtr position);
         [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
         private delegate void ResizeCompletedEventCallbackType(IntPtr window, IntPtr size);
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
```
thank you for your fix.
but could you change this as `.Cdecl` ?

jaehyun0cho replied:
Thank you for your comment!
I updated the code as you requested.

dongsug-song replied:
thank you for fix
